bpf/bpf_sock.c

```diff
@@ -414,6 +418,10 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 
 	if (lb4_svc_is_affinity(svc) && !backend_from_affinity)
 		lb4_update_affinity_by_netns(svc, &id, backend_id);
+
+	send_trace_sock_notify4(ctx_full, XLATE_POST_DIRECTION_FWD, backend->address,
+				bpf_ntohs(backend->port));
+
```

Comment:
From trace PoV, what is the rationale for XLATE_PRE_DIRECTION_FWD and XLATE_POST_DIRECTION_FWD compared to just having a XLATE_DIRECTION_FWD hook which is done after each point where we rewrite ctx->user_ip and port?

Meaning, adding XLATE_DIRECTION_FWD with svc vip/port plus backend ip/port info that we pass to send_trace_sock_notify4() below before the return 0, and also for the L7 case where we write (which seems missing here?).

And then having XLATE_ERROR_... tracepoints where we get more visibility in some of the error cases which are proper errors, like failing to update sock4_update_revnat etc. Wdyt?

Reply:
Discussed offline, posting here for posterity.

The intent behind having pre and post service translation trace points is to enable users to track the logic separately. For early returns, we already have error metrics for visibility. 

The error codes are out of the scope of the PR, but I'm happy to discuss this so that we can extend this in follow-ups if/as needed.